<!DOCTYPE html>
<html lang="ca">
<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Gestió d'Errors Avançada</title>

    <link rel="alternate" href="https://campusempresa.com/mod/swift/05-02-advanced-error-handling" hreflang="es" />
	<link rel="alternate" href="https://campusempresa.cat/mod/swift/05-02-advanced-error-handling" hreflang="ca" />
	<link rel="alternate" href="https://enterprisecampus.net/mod/swift/05-02-advanced-error-handling" hreflang="en" />
    
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/css/bootstrap.min.css" rel="stylesheet">
	<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons/font/bootstrap-icons.css">
	<link href="/css/site.css" rel="stylesheet">
	
    <script src="https://code.jquery.com/jquery-3.7.1.min.js"></script>
  	<script type="text/javascript" src="/js/math_init.js"></script>
  	<script type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/startup.js"></script>
  	<script type="text/javascript" src="/js/cookie.js"></script>
  	<script type="text/javascript" src="/js/main.js"></script>
</head>

<body >
    <div id="content">
		<div id="header" class="container-xxl">
	<div class="row">
		<div class="col-12 col-md-8 p-2 p-md-0">
			<h1 class="m-0 p-0">
				<a href="/"><img src="/img/logo_header.png"></a>
			</h1>
		</div>
		<div class="col-12 col-md-4 p-2 p-md-0 text-end">
			<h2 id="main_title"><cite>Construint la societat d'avui i del demà</cite></h2>
			<h3 id="main_subtitle"></h3>
		</div>
	</div>
</div>
<div class="container-xxl" style="margin-top: -1em;">
	<div class="row">
		<div class="col-12 p-2 p-md-0 m-0 text-end">
							<a href="https://enterprisecampus.net/mod/swift/05-02-advanced-error-handling" class="px-2">EN</a></b>
				|
				<a href="https://campusempresa.com/mod/swift/05-02-advanced-error-handling" class="px-2">ES</a></b>
				|
				<b class="px-2">CA</b>
											</div>
	</div>
</div>
   <div class="top-bar container-fluid">
	<div class="container-xxl">
		<div class="row">
			<div class="col" id="left_menu">
				<a href="/objective">El Projecte</a>
				<a href="/about">Sobre nosaltres</a>
				<a href="/contribute">Contribuir</a>
				<a href="/donate">Donacions</a>
				<a href="/licence">Llicència</a>
			</div>
		</div>
	</div>
   </div>

<div class="container-xxl" id="main_content">
	<div class="row">
		<div class="col-12 col-lg-8">
			<div id="nav1" class="navigation"></div>
			<div id="inner_content">
								<div class='row navigation'>
	<div class='col-2'>
					<a href='05-01-generics' title="Genèrics">&#x25C4;Anterior</a>
			</div>
	<div class='col-8 text-center'>
					<a href="./"><h2 style="text-decoration:underline">Gestió d'Errors Avançada</h2></a>
			</div>
	<div class='col-2 text-end'>
					<a href='05-03-memory-management' title="Gestió de Memòria">Següent &#x25BA;</a>
			</div>
</div>
<div class='content'><p>En aquest tema, explorarem tècniques avançades per gestionar errors en Swift. La gestió d'errors és una part crucial del desenvolupament de programari, ja que permet als desenvolupadors manejar situacions inesperades de manera controlada i segura.</p>
</div><h1><p>Continguts</p>
</h1>
<div class='content'><ol>
<li><a href="#revisió-de-la-gestió-derrors-bàsica">Revisió de la Gestió d'Errors Bàsica</a></li>
<li><a href="#errors-personalitzats">Errors Personalitzats</a></li>
<li><a href="#propagació-derrors">Propagació d'Errors</a></li>
<li><a href="#captura-i-maneig-derrors">Captura i Maneig d'Errors</a></li>
<li><a href="#errors-i-funcions-asíncrones">Errors i Funcions Asíncrones</a></li>
<li><a href="#exercicis-pràctics">Exercicis Pràctics</a></li>
</ol>
</div><h1><p>Revisió de la Gestió d'Errors Bàsica</p>
</h1>
<div class='content'><p>Abans d'endinsar-nos en tècniques avançades, fem una breu revisió de la gestió d'errors bàsica en Swift.</p>
</div><h2><p>Declaració d'Errors</p>
</h2>
<div class='content'><p>En Swift, els errors es declaren utilitzant el protocol <code>Error</code>. Aquí teniu un exemple:</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZW51bSBGaWxlRXJyb3I6IEVycm9yIHsKICAgIGNhc2UgZmlsZU5vdEZvdW5kCiAgICBjYXNlIHVucmVhZGFibGUKICAgIGNhc2UgZW5jb2RpbmdGYWlsZWQKfQ=="))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>enum FileError: Error {
    case fileNotFound
    case unreadable
    case encodingFailed
}</pre></div><div class='content'></div><h2><p>Llençar Errors</p>
</h2>
<div class='content'><p>Per llençar un error, utilitzem la paraula clau <code>throw</code>:</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZnVuYyByZWFkRmlsZShhdCBwYXRoOiBTdHJpbmcpIHRocm93cyAtPiBTdHJpbmcgewogICAgLy8gU2ltdWxhY2nDsyBkJ3VuIGVycm9yCiAgICB0aHJvdyBGaWxlRXJyb3IuZmlsZU5vdEZvdW5kCn0="))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>func readFile(at path: String) throws -&gt; String {
    // Simulaci&oacute; d'un error
    throw FileError.fileNotFound
}</pre></div><div class='content'></div><h2><p>Captura d'Errors</p>
</h2>
<div class='content'><p>Els errors es capturen utilitzant una estructura <code>do-catch</code>:</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZG8gewogICAgbGV0IGNvbnRlbnQgPSB0cnkgcmVhZEZpbGUoYXQ6ICJwYXRoL3RvL2ZpbGUiKQogICAgcHJpbnQoY29udGVudCkKfSBjYXRjaCBGaWxlRXJyb3IuZmlsZU5vdEZvdW5kIHsKICAgIHByaW50KCJFbCBmaXR4ZXIgbm8gcydoYSB0cm9iYXQuIikKfSBjYXRjaCB7CiAgICBwcmludCgiUydoYSBwcm9kdcOvdCB1biBlcnJvcjogXChlcnJvcikuIikKfQ=="))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>do {
    let content = try readFile(at: &quot;path/to/file&quot;)
    print(content)
} catch FileError.fileNotFound {
    print(&quot;El fitxer no s'ha trobat.&quot;)
} catch {
    print(&quot;S'ha produ&iuml;t un error: \(error).&quot;)
}</pre></div><div class='content'></div><h1><p>Errors Personalitzats</p>
</h1>
<div class='content'><p>Crear errors personalitzats ens permet proporcionar informació més detallada sobre els errors que es produeixen.</p>
</div><h2><p>Exemple d'Error Personalitzat</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZW51bSBOZXR3b3JrRXJyb3I6IEVycm9yIHsKICAgIGNhc2UgYmFkVVJMCiAgICBjYXNlIHJlcXVlc3RGYWlsZWQoZGVzY3JpcHRpb246IFN0cmluZykKICAgIGNhc2UgdW5rbm93bgp9"))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>enum NetworkError: Error {
    case badURL
    case requestFailed(description: String)
    case unknown
}</pre></div><div class='content'></div><h2><p>Utilització d'Errors Personalitzats</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZnVuYyBmZXRjaERhdGEoZnJvbSB1cmxTdHJpbmc6IFN0cmluZykgdGhyb3dzIC0+IERhdGEgewogICAgZ3VhcmQgbGV0IHVybCA9IFVSTChzdHJpbmc6IHVybFN0cmluZykgZWxzZSB7CiAgICAgICAgdGhyb3cgTmV0d29ya0Vycm9yLmJhZFVSTAogICAgfQogICAgCiAgICAvLyBTaW11bGFjacOzIGQndW4gZXJyb3IgZGUgc29swrdsaWNpdHVkCiAgICB0aHJvdyBOZXR3b3JrRXJyb3IucmVxdWVzdEZhaWxlZChkZXNjcmlwdGlvbjogIkxhIHNvbMK3bGljaXR1ZCBoYSBmYWxsYXQgYW1iIHVuIGNvZGkgZCdlc3RhdCA0MDQuIikKfQ=="))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>func fetchData(from urlString: String) throws -&gt; Data {
    guard let url = URL(string: urlString) else {
        throw NetworkError.badURL
    }
    
    // Simulaci&oacute; d'un error de sol&middot;licitud
    throw NetworkError.requestFailed(description: &quot;La sol&middot;licitud ha fallat amb un codi d'estat 404.&quot;)
}</pre></div><div class='content'></div><h1><p>Propagació d'Errors</p>
</h1>
<div class='content'><p>La propagació d'errors permet que una funció que llença errors pugui ser cridada des d'una altra funció que també llença errors.</p>
</div><h2><p>Exemple de Propagació d'Errors</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZnVuYyBwZXJmb3JtTmV0d29ya1JlcXVlc3QoKSB0aHJvd3MgewogICAgdHJ5IGZldGNoRGF0YShmcm9tOiAiaHR0cHM6Ly9leGFtcGxlLmNvbSIpCn0KCmRvIHsKICAgIHRyeSBwZXJmb3JtTmV0d29ya1JlcXVlc3QoKQp9IGNhdGNoIE5ldHdvcmtFcnJvci5iYWRVUkwgewogICAgcHJpbnQoIlVSTCBpbmNvcnJlY3RhLiIpCn0gY2F0Y2ggTmV0d29ya0Vycm9yLnJlcXVlc3RGYWlsZWQobGV0IGRlc2NyaXB0aW9uKSB7CiAgICBwcmludCgiRXJyb3IgZGUgc29swrdsaWNpdHVkOiBcKGRlc2NyaXB0aW9uKS4iKQp9IGNhdGNoIHsKICAgIHByaW50KCJTJ2hhIHByb2R1w690IHVuIGVycm9yIGRlc2NvbmVndXQuIikKfQ=="))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>func performNetworkRequest() throws {
    try fetchData(from: &quot;https://example.com&quot;)
}

do {
    try performNetworkRequest()
} catch NetworkError.badURL {
    print(&quot;URL incorrecta.&quot;)
} catch NetworkError.requestFailed(let description) {
    print(&quot;Error de sol&middot;licitud: \(description).&quot;)
} catch {
    print(&quot;S'ha produ&iuml;t un error desconegut.&quot;)
}</pre></div><div class='content'></div><h1><p>Captura i Maneig d'Errors</p>
</h1>
<div class='content'></div><h2><p>Captura d'Errors Específics</p>
</h2>
<div class='content'><p>Podem capturar errors específics i manejar-los de manera diferent:</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZG8gewogICAgdHJ5IHBlcmZvcm1OZXR3b3JrUmVxdWVzdCgpCn0gY2F0Y2ggTmV0d29ya0Vycm9yLmJhZFVSTCB7CiAgICBwcmludCgiVVJMIGluY29ycmVjdGEuIikKfSBjYXRjaCBOZXR3b3JrRXJyb3IucmVxdWVzdEZhaWxlZChsZXQgZGVzY3JpcHRpb24pIHsKICAgIHByaW50KCJFcnJvciBkZSBzb2zCt2xpY2l0dWQ6IFwoZGVzY3JpcHRpb24pLiIpCn0gY2F0Y2ggewogICAgcHJpbnQoIlMnaGEgcHJvZHXDr3QgdW4gZXJyb3IgZGVzY29uZWd1dC4iKQp9"))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>do {
    try performNetworkRequest()
} catch NetworkError.badURL {
    print(&quot;URL incorrecta.&quot;)
} catch NetworkError.requestFailed(let description) {
    print(&quot;Error de sol&middot;licitud: \(description).&quot;)
} catch {
    print(&quot;S'ha produ&iuml;t un error desconegut.&quot;)
}</pre></div><div class='content'></div><h2><p>Captura d'Errors Generals</p>
</h2>
<div class='content'><p>Podem capturar errors generals utilitzant un bloc <code>catch</code> sense especificar el tipus d'error:</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZG8gewogICAgdHJ5IHBlcmZvcm1OZXR3b3JrUmVxdWVzdCgpCn0gY2F0Y2ggewogICAgcHJpbnQoIlMnaGEgcHJvZHXDr3QgdW4gZXJyb3I6IFwoZXJyb3IpLiIpCn0="))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>do {
    try performNetworkRequest()
} catch {
    print(&quot;S'ha produ&iuml;t un error: \(error).&quot;)
}</pre></div><div class='content'></div><h1><p>Errors i Funcions Asíncrones</p>
</h1>
<div class='content'><p>Quan treballem amb funcions asíncrones, la gestió d'errors es complica una mica. Swift proporciona mecanismes per manejar errors en contextos asíncrons.</p>
</div><h2><p>Exemple amb <code>async</code> i <code>await</code></p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZW51bSBBc3luY0Vycm9yOiBFcnJvciB7CiAgICBjYXNlIHRpbWVvdXQKfQoKZnVuYyBmZXRjaERhdGFBc3luYygpIGFzeW5jIHRocm93cyAtPiBEYXRhIHsKICAgIC8vIFNpbXVsYWNpw7MgZCd1biBlcnJvciBhc8OtbmNyb24KICAgIHRocm93IEFzeW5jRXJyb3IudGltZW91dAp9CgpmdW5jIHBlcmZvcm1Bc3luY1JlcXVlc3QoKSBhc3luYyB7CiAgICBkbyB7CiAgICAgICAgbGV0IGRhdGEgPSB0cnkgYXdhaXQgZmV0Y2hEYXRhQXN5bmMoKQogICAgICAgIHByaW50KCJEYWRlcyByZWJ1ZGVzOiBcKGRhdGEpIikKICAgIH0gY2F0Y2ggQXN5bmNFcnJvci50aW1lb3V0IHsKICAgICAgICBwcmludCgiTGEgc29swrdsaWNpdHVkIGhhIHN1cGVyYXQgZWwgdGVtcHMgZCdlc3BlcmEuIikKICAgIH0gY2F0Y2ggewogICAgICAgIHByaW50KCJTJ2hhIHByb2R1w690IHVuIGVycm9yOiBcKGVycm9yKS4iKQogICAgfQp9"))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>enum AsyncError: Error {
    case timeout
}

func fetchDataAsync() async throws -&gt; Data {
    // Simulaci&oacute; d'un error as&iacute;ncron
    throw AsyncError.timeout
}

func performAsyncRequest() async {
    do {
        let data = try await fetchDataAsync()
        print(&quot;Dades rebudes: \(data)&quot;)
    } catch AsyncError.timeout {
        print(&quot;La sol&middot;licitud ha superat el temps d'espera.&quot;)
    } catch {
        print(&quot;S'ha produ&iuml;t un error: \(error).&quot;)
    }
}</pre></div><div class='content'></div><h1><p>Exercicis Pràctics</p>
</h1>
<div class='content'></div><h2><p>Exercici 1: Crear un Error Personalitzat</p>
</h2>
<div class='content'><ol>
<li>Defineix un error personalitzat anomenat <code>DatabaseError</code> amb els casos <code>connectionFailed</code> i <code>dataNotFound</code>.</li>
<li>Escriu una funció <code>fetchDataFromDatabase</code> que llença aquests errors.</li>
<li>Escriu un bloc <code>do-catch</code> per capturar i manejar aquests errors.</li>
</ol>
</div><h2><p>Solució</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZW51bSBEYXRhYmFzZUVycm9yOiBFcnJvciB7CiAgICBjYXNlIGNvbm5lY3Rpb25GYWlsZWQKICAgIGNhc2UgZGF0YU5vdEZvdW5kCn0KCmZ1bmMgZmV0Y2hEYXRhRnJvbURhdGFiYXNlKCkgdGhyb3dzIC0+IFN0cmluZyB7CiAgICAvLyBTaW11bGFjacOzIGQndW4gZXJyb3IKICAgIHRocm93IERhdGFiYXNlRXJyb3IuY29ubmVjdGlvbkZhaWxlZAp9CgpkbyB7CiAgICBsZXQgZGF0YSA9IHRyeSBmZXRjaERhdGFGcm9tRGF0YWJhc2UoKQogICAgcHJpbnQoIkRhZGVzIHJlYnVkZXM6IFwoZGF0YSkiKQp9IGNhdGNoIERhdGFiYXNlRXJyb3IuY29ubmVjdGlvbkZhaWxlZCB7CiAgICBwcmludCgiTGEgY29ubmV4acOzIGEgbGEgYmFzZSBkZSBkYWRlcyBoYSBmYWxsYXQuIikKfSBjYXRjaCBEYXRhYmFzZUVycm9yLmRhdGFOb3RGb3VuZCB7CiAgICBwcmludCgiTGVzIGRhZGVzIG5vIHMnaGFuIHRyb2JhdC4iKQp9IGNhdGNoIHsKICAgIHByaW50KCJTJ2hhIHByb2R1w690IHVuIGVycm9yOiBcKGVycm9yKS4iKQp9"))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>enum DatabaseError: Error {
    case connectionFailed
    case dataNotFound
}

func fetchDataFromDatabase() throws -&gt; String {
    // Simulaci&oacute; d'un error
    throw DatabaseError.connectionFailed
}

do {
    let data = try fetchDataFromDatabase()
    print(&quot;Dades rebudes: \(data)&quot;)
} catch DatabaseError.connectionFailed {
    print(&quot;La connexi&oacute; a la base de dades ha fallat.&quot;)
} catch DatabaseError.dataNotFound {
    print(&quot;Les dades no s'han trobat.&quot;)
} catch {
    print(&quot;S'ha produ&iuml;t un error: \(error).&quot;)
}</pre></div><div class='content'></div><h2><p>Exercici 2: Manejar Errors en Funcions Asíncrones</p>
</h2>
<div class='content'><ol>
<li>Defineix una funció asíncrona <code>loadData</code> que llença un error <code>AsyncError</code>.</li>
<li>Escriu una funció <code>performLoadData</code> que crida <code>loadData</code> utilitzant <code>await</code> i maneja els errors.</li>
</ol>
</div><h2><p>Solució</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("ZW51bSBBc3luY0Vycm9yOiBFcnJvciB7CiAgICBjYXNlIHRpbWVvdXQKfQoKZnVuYyBsb2FkRGF0YSgpIGFzeW5jIHRocm93cyAtPiBTdHJpbmcgewogICAgLy8gU2ltdWxhY2nDsyBkJ3VuIGVycm9yIGFzw61uY3JvbgogICAgdGhyb3cgQXN5bmNFcnJvci50aW1lb3V0Cn0KCmZ1bmMgcGVyZm9ybUxvYWREYXRhKCkgYXN5bmMgewogICAgZG8gewogICAgICAgIGxldCBkYXRhID0gdHJ5IGF3YWl0IGxvYWREYXRhKCkKICAgICAgICBwcmludCgiRGFkZXMgY2FycmVnYWRlczogXChkYXRhKSIpCiAgICB9IGNhdGNoIEFzeW5jRXJyb3IudGltZW91dCB7CiAgICAgICAgcHJpbnQoIkxhIGPDoHJyZWdhIGRlIGRhZGVzIGhhIHN1cGVyYXQgZWwgdGVtcHMgZCdlc3BlcmEuIikKICAgIH0gY2F0Y2ggewogICAgICAgIHByaW50KCJTJ2hhIHByb2R1w690IHVuIGVycm9yOiBcKGVycm9yKS4iKQogICAgfQp9"))));alert("Copiat!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>enum AsyncError: Error {
    case timeout
}

func loadData() async throws -&gt; String {
    // Simulaci&oacute; d'un error as&iacute;ncron
    throw AsyncError.timeout
}

func performLoadData() async {
    do {
        let data = try await loadData()
        print(&quot;Dades carregades: \(data)&quot;)
    } catch AsyncError.timeout {
        print(&quot;La c&agrave;rrega de dades ha superat el temps d'espera.&quot;)
    } catch {
        print(&quot;S'ha produ&iuml;t un error: \(error).&quot;)
    }
}</pre></div><div class='content'></div><h1><p>Conclusió</p>
</h1>
<div class='content'><p>En aquesta secció, hem explorat tècniques avançades per gestionar errors en Swift, incloent la creació d'errors personalitzats, la propagació d'errors, la captura i maneig d'errors, i la gestió d'errors en funcions asíncrones. La gestió d'errors és una habilitat essencial per a qualsevol desenvolupador, ja que permet crear aplicacions més robustes i fiables. En el proper tema, ens endinsarem en la gestió de memòria en Swift.</p>
</div><div class='row navigation'>
	<div class='col-2'>
					<a href='05-01-generics' title="Genèrics">&#x25C4;Anterior</a>
			</div>
	<div class='col-8 text-center'>
			</div>
	<div class='col-2 text-end'>
					<a href='05-03-memory-management' title="Gestió de Memòria">Següent &#x25BA;</a>
			</div>
</div>

			</div>
		</div>
		<div class="col-12 col-lg-4 publi" id="div_publi">
			<h1>Publicitat</h1>
			<p>Aquest espai està destinat a publicitat.</p>
			<p>Si vols ser patrocinador, contacta amb nosaltres per incloure enllaços en aquesta zona: <a href='mailto:admin@campusempresa.cat'>admin@campusempresa.cat</a></p>
			<p>Gràcies per col·laborar!</p>
		</div>
	</div>
</div>

   <div class="container-xxl my-3">
	<div class="row">
		<div class="col">
			<footer>&copy; Copyright 2024. Tots els drets reservats</footer>
		</div>
	</div>
</div>	

<div id="cookies_adv" style="display:none;">
	Fem servir galetes per millorar la teva experiència d'ús i oferir continguts adaptats als teus interessos
    <a href="#" id="btn_accept_cookies" class="button">Acceptar</a>
    <a href="/cookies">Més informació</a>
</div>	

	</div>    
</body>
</html>
